### ISSUE TITLE
Feature Request: Self hosting

### ISSUE THREAD

**Issue by patha454** [Textualize/textual-web#29]
The idea of serving Textual apps over the internet is interesting. I'm not comfortable with the idea of requiring a proxy over the `textual-web.io` gateway for the obvious security reasons - exposing a terminal to the public internet...

Are there plans for for a feature to open a server on the local machine, and use our own web infrastructure, firewalls, and authentication?

**Comment by willmcgugan** [comment]
Authentication and end-to-end encryption are planned.

We may offer an "on-prem" solution at some point. But likely not until the SaaS product has been refined.

**Comment by sporqist** [comment]
@patha454 Using `--environment local` will make your app connect to a local endpoint instead of https://textual-web.io. This will fail as textual-web doesn't actually come with a backend. It will fail with this warning:
```
WARNING  Unable to connect to Ganglion server. Will reattempt
                    connection soon.
```

I tried researching Ganglion server to learn more about the backend but couldn't find anything about it. Seems to be closed source.(?)

@davep I understand you want to deliver a SaaS product here but how would I go about developing a textual-web app in a secure way? Actually, I was stumped when I realised, the default behavior of `textual-web` was to expose an app running on my own machine to the internet without any warning.

**Comment by sporqist** [comment]
@davep Sorry about that. I saw your link to this and made the wrong assumption. Edited my comment.

**Comment by willmcgugan** [comment]
The server is proprietary at the moment. We have no short-term plans for a local version, but it could happen.

What concerns do you have about running your app in a "secure way"? You can run your app without textual-web, and only deploy it when its ready.

**Comment by panki27** [comment]
@willmcgugan my understanding of "deploying an app" is that I install things on a server that I control.
If something about the app goes down, that's should be on me because I messed up. 

I do not want to run something that is essentially a reverse tunnel to a third party site just to serve my local application as an HTTP site. I guess an alternative could be setting up something like `shellinabox` that is configured to run the Python file.

**Comment by patha454** [comment]
Quick note: The author has since released https://github.com/Textualize/textual-serve, which is a locally hosted version.

**Comment by alexander-potemkin** [comment]
@patha454 , thanks! Do you believe / have some experience, if it's production ready?

**Comment by patha454** [comment]
@alexander-potemkin 

Depends on your use case. The author (Will) is quite clear `textual-serve` is in it's early days - indeed it only appeared on public Github two weeks ago.

As for production ready.. that depends on your use case:

- It should be self-evident Will writes rather high quality code.
- Textual-serve is short and trivial. The brief, simple code presents few changes for major bugs to creep in.
  - A few hundred lines of Python for starting the app and communicating with Textual.
  - A obfuscated JS blob for rendering & encoding/decoding data from text.
- The obfuscated JS is presumably so Will has the monetize parts of Textual - he's stated he wants to produce a SaaS offering based around Textual-web.
  - I have no reason to suspect the blob contains malware, but it's a theoretical possibility with any code we can't easily audit - audit and do risk analysis commiserate with your use case.
  - Textual-serve has no authentication built in at this point.
  - Program instances are separate subprocess, with no other isolation: You could have the noisy-neighbors problem, and
 
Textual-serve is brand new, and Textual is still young by industrial standards. Some teething problems and "early-adopter tax" should be expected.

As with any open source code, do your own analysis, auditing, and pen-testing in line with how critical your use case is.

In my opinion (and very limited experience - we're still in the analysis phase), Textual-serve would be suitable for in-house projects on a private network, with some standard precautions.

You may want to wait-for, or add your own, authentication if your app or network configuration arn't already doing that for you.

I wouldn't be comfortable using textual-serve on the public internet quite yet (I need more time to audit and trial everything. We're fairly risk-adverse, your organization may not be as pedantic as ours.) I don't see any reason why it wouldn't get there: Textual is fairly reliable, Textual-serve is mostly trivial, and Textual-serve uses a fairly popular HTTP router.

I wouldn't be comfortable using Textual-serve to touch any sensitive private information - but that's just the being wary of anything new and unproven for sensitive use case.

In summary, Textual-serve (and the underlying Textual), are solid, but new tools. I'd be comfortable using them for low-risk production situations. As always, do your own audit and risk analysis; decide how sensitive your use case is and how comfortable you are using relatively immature tools.

**Comment by willmcgugan** [comment]
@patha454 I'm deleting your comment because it is misinformation. Please do not report assumptions as facts.

@alexander-potemkin Textual-serve is indeed production ready. It will grow a few more features in the near future to give Textual apps greater parity with conventional web apps.

**Comment by alexander-potemkin** [comment]
@willmcgugan , I understand you might delete my comment as well, but I see nothing wrong in what @patha454 commented earlier.

As a matter of fact, what I've seen was:
- you create high-quality code
- web part of it is not well tested as it might be in some use cases, which is I assume, is to compare with nginx or Django code bases, for example
- there is obfuscated JS part - which is, from my personal perspective, feels like a genius solution to the demand you have - to provide self-hosting option
- you might not want to serve textual-web to provide, for example, banking experience (see point on the code base comparison), but it seems to be Ok for most of the use cases.

I understand that I might be missing something, but for me, it's all about compliments to you and to the product from the perspective of the guy with an experience in very demanding environments (f.ex.: banks, insurance, medical, etc).

And, in any case, thanks for that self-hosted web solution! It's a really nice one!
